src/orange/detail/OrientedBoundingZone.hh

```diff
@@ -0,0 +1,283 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/OrientedBoundingZone.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/NumericLimits.hh"
+#include "orange/OrangeData.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/transform/TransformVisitor.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Oriented bounding zone for saftey distance calculations.
+ *
+ * Here, the bounding zone is defined by an inner and outer bounding box
+ * (specified via half-widths) transformed by the same transformation;
+ * pressumably the transformation of the volume they correspond to. The
+ * safety distance can be approximated as the minimum distance to the inner or
+ * outer box. For points lying between the inner and outer boxes, the safety
+ * distance is zero.
+ */
+class OrientedBoundingZone
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    template<class T>
+    using StorageItems
+        = Collection<T, Ownership::const_reference, MemSpace::native>;
+
+    using ObzReal3 = OrientedBoundingZoneRecord::Real3;
+    using ObzReal3Id = OrientedBoundingZoneRecord::Real3Id;
+
+    struct Storage
+    {
+        StorageItems<ObzReal3> const* half_widths;
+        StorageItems<TransformRecord> const* transforms;
+        StorageItems<real_type> const* reals;
+    };
+    //!@}
+
+  public:
+    // Construct from inner/outer half-widths and a corresponding transform
+    inline CELER_FUNCTION
+    OrientedBoundingZone(OrientedBoundingZoneRecord const* obz_record,
+                         Storage const* storage);
+
+    // Calculate the safety distance for any position inside the outer box
+    inline CELER_FUNCTION real_type safety_distance_inside(Real3 pos);
+
+    // Calculate the safety distance for any position outside the inner box
+    inline CELER_FUNCTION real_type safety_distance_outside(Real3 pos);
+
+    // Determine the sense of position with respect to the bounding zone
+    inline CELER_FUNCTION SignedSense calc_sense(Real3 const& pos);
+
+  private:
+    // >> DATA
+    OrientedBoundingZoneRecord const* obz_record_;
+    Storage const* storage_;
```

Comment:
Since `Storage` here is actually pointers and not a Collection, I think it would be better to rename as `Input` or `StoragePointers` and you definitely should store it by value.